File: tests/python/tests-python-playwright/screenshot_helper.py
"""PNG capture/compare — mirrors Java ScreenshotHelper (LAYERS.md Mock and screenshot)."""

from __future__ import annotations

import io
import os
import sys
from pathlib import Path

import allure
from PIL import Image
from config import load_config

_ROOT = Path(__file__).resolve().parent
_DIFF_DIR = _ROOT / "build" / "screenshot-diff"
_DIFF_HIGHLIGHT = (255, 0, 255)
_SIZE_MISMATCH = (255, 0, 0)


def capture_and_compare(locator, area: str, viewport: int, attachment_name: str) -> None:
    wait_for_stable_layout(locator.page)
    actual = locator.screenshot()
    label = f"{area}/{viewport}"
    screenshot_path = screenshot_file_path(area, viewport)
    present = screenshot_path.is_file()

    if _should_update():
        with allure.step(f"Update screenshot: {attachment_name}"):
            _attach_update_mode(attachment_name, actual, present, area, viewport)
        _write_screenshot(screenshot_path, actual)
        return

    if not present:
        with allure.step(f"Missing screenshot: {attachment_name}"):
            _attach_png(f"{attachment_name}-actual-unmatched", actual)
        resource = screenshot_resource_path(area, viewport)
        raise AssertionError(
            f"Screenshot missing for {label}. Commit PNG to src/test/resources/{resource} "
            "or run with UPDATE_SCREENSHOTS=true"
        )

    expected = screenshot_path.read_bytes()
    passed, diff_png, message = compare_images(expected, actual, label)
    with allure.step(f"Compare screenshot: {attachment_name}"):
        if passed:
            _attach_png(attachment_name, actual)
            return
        _attach_png(f"{attachment_name}-expected", expected)
        _attach_png(f"{attachment_name}-actual", actual)
        _attach_png(f"{attachment_name}-diff", diff_png)
        _save_fail_artifacts(label, actual, diff_png)
        raise AssertionError(message)


def wait_for_stable_layout(page) -> None:
    page.evaluate(
        """() => Promise.all([
          document.fonts.ready,
          new Promise(r => requestAnimationFrame(() => requestAnimationFrame(r)))
        ])"""
    )


def screenshot_mode(env: str | None = None) -> str:
    if env is None:
        env = os.environ.get("STAND") or os.environ.get("ENV") or ""
    key = (env or "").strip()
    if key == "mock":
        return "mock"
    if key == "stage":
        return "stage"
    if key in {"prod", "ci", ""}:
        return "prod"
    raise ValueError(f"screenshot folder: unknown env '{key}' (use mock, stage, prod, or ci)")


def screenshot_os() -> str:
    override = os.environ.get("SCREENSHOT_OS")
    raw = override.strip() if override and override.strip() else _os_family()
    return _map_screenshot_os(raw)


def screenshot_browser_folder() -> str:
    return f"{screenshot_browser()}-{screenshot_browser_major()}"


def screenshot_browser() -> str:
    override = os.environ.get("SCREENSHOT_BROWSER")
    if override and override.strip():
        return override.strip().lower()
    return "chrome"


def screenshot_browser_major() -> str:
    return pinned_chrome_version().split(".", 1)[0]


def pinned_chrome_version() -> str:
    override = os.environ.get("CHROME_FOR_TESTING_VERSION", "").strip()
    if override:
        return override
    pin = _ROOT / "chrome-for-testing.properties"
    if not pin.is_file():
        raise RuntimeError(f"chrome-for-testing.properties is missing: {pin}")
    for line in pin.read_text(encoding="utf-8").splitlines():
        stripped = line.strip()
        if stripped.startswith("version="):
            value = stripped.split("=", 1)[1].strip()
            if value:
                return value
    raise RuntimeError("No version= entry in chrome-for-testing.properties")


def screenshot_file_path(area: str, viewport: int) -> Path:
    cfg = load_config()
    screenshots_dir = (cfg.screenshots_dir or "screenshots").replace("\\", "/").strip("/")
    return (
        _ROOT
        / "src"
        / "test"
        / "resources"
        / screenshots_dir
        / screenshot_mode(cfg.stand)
        / screenshot_os()
        / screenshot_browser_folder()
        / area
        / f"{viewport}.png"
    )


def screenshot_resource_path(area: str, viewport: int) -> str:
    cfg = load_config()
    screenshots_dir = (cfg.screenshots_dir or "screenshots").replace("\\", "/").strip("/")
    return (
        f"{screenshots_dir}/{screenshot_mode(cfg.stand)}/{screenshot_os()}/"
        f"{screenshot_browser_folder()}/{area}/{viewport}.png"
    )


def compare_images(expected_bytes: bytes, actual_bytes: bytes, label: str) -> tuple[bool, bytes, str | None]:
    expected = _read_image(expected_bytes)
    actual = _read_image(actual_bytes)
    diff_png = _create_diff_png(expected, actual)
    if expected.size != actual.size:
        return (
            False,
            diff_png,
            "Screenshot size changed for %s: expected %dx%d, actual %dx%d"
            % (label, expected.width, expected.height, actual.width, actual.height),
        )

    width, height = expected.size
    exp_bytes = expected.tobytes()
    act_bytes = actual.tobytes()
    diff_pixels = 0
    for i in range(0, len(exp_bytes), 3):
        if exp_bytes[i : i + 3] != act_bytes[i : i + 3]:
            diff_pixels += 1
    total = width * height
    max_diff = load_config().screenshot_diff_threshold
    ratio = diff_pixels / total if total else 1.0
    if ratio > max_diff:
        return (
            False,
            diff_png,
            "Screenshot diff too high for %s: %.2f%% > %.2f%%" % (label, ratio * 100, max_diff * 100),
        )
    return True, diff_png, None


def _should_update() -> bool:
    return load_config().update_screenshots


def _os_family() -> str:
    if sys.platform == "darwin":
        return "darwin"
    if sys.platform.startswith("win"):
        return "win32"
    return "linux"


def _map_screenshot_os(raw: str) -> str:
    key = raw.lower()
    if key in {"darwin", "macos"} or key.startswith("mac"):
        return "macos"
    if key in {"win32", "windows"} or key.startswith("win"):
        return "windows"
    if key == "linux" or "linux" in key:
        return "linux"
    return key if key else "linux"


def _attach_update_mode(
    attachment_name: str, actual: bytes, present: bool, area: str, viewport: int
) -> None:
    if present:
        _attach_png(f"{attachment_name}-screenshot-old", screenshot_file_path(area, viewport).read_bytes())
        _attach_png(f"{attachment_name}-screenshot-new", actual)
        return
    _attach_png(f"{attachment_name}-screenshot-new", actual)


def _attach_png(name: str, png: bytes) -> None:
    allure.attach(png, name=name, attachment_type=allure.attachment_type.PNG)


def _write_screenshot(path: Path, png: bytes) -> None:
    path.parent.mkdir(parents=True, exist_ok=True)
    path.write_bytes(png)


def _save_fail_artifacts(label: str, actual: bytes, diff: bytes) -> None:
    try:
        _DIFF_DIR.mkdir(parents=True, exist_ok=True)
        prefix = label.replace("/", "_")
        (_DIFF_DIR / f"{prefix}-actual.png").write_bytes(actual)
        (_DIFF_DIR / f"{prefix}-diff.png").write_bytes(diff)
    except OSError:
        pass


def _read_image(data: bytes) -> Image.Image:
    image = Image.open(io.BytesIO(data))
    if image is None:
        raise OSError("Unsupported screenshot format")
    return image.convert("RGB")


def _create_diff_png(expected: Image.Image, actual: Image.Image) -> bytes:
    exp_w, exp_h = expected.size
    act_w, act_h = actual.size
    width = max(exp_w, act_w)
    height = max(exp_h, act_h)
    exp = expected.load()
    act = actual.load()
    diff = Image.new("RGB", (width, height), _SIZE_MISMATCH)
    px = diff.load()
    for y in range(height):
        for x in range(width):
            in_expected = x < exp_w and y < exp_h
            in_actual = x < act_w and y < act_h
            if in_expected and in_actual:
                expected_rgb = exp[x, y]
                if expected_rgb == act[x, y]:
                    px[x, y] = _dim_rgb(expected_rgb)
                else:
                    px[x, y] = _DIFF_HIGHLIGHT
    out = io.BytesIO()
    diff.save(out, format="PNG")
    return out.getvalue()


def _dim_rgb(rgb: tuple[int, int, int]) -> tuple[int, int, int]:
    dim = (rgb[0] + rgb[1] + rgb[2]) // 9
    return dim, dim, dim

File: tests/csharp/tests-csharp-xunit-api_request-playwright/Helpers/ScreenshotHelper.cs
using Allure.Net.Commons;
using Config;
using Microsoft.Playwright;
using SixLabors.ImageSharp;
using SixLabors.ImageSharp.PixelFormats;
using Xunit.Sdk;

namespace Helpers;

public static class ScreenshotHelper
{
    private static readonly string DiffDir = Path.Combine("screenshot-diff");
    private static readonly Rgba32 DiffHighlight = new(255, 0, 255);
    private static readonly Rgba32 SizeMismatch = new(255, 0, 0);

    public static void CaptureAndCompare(ILocator element, string area, int viewport, string attachmentName)
    {
        Pw.Run(element.WaitForAsync());
        var actual = Pw.Run(element.ScreenshotAsync(new LocatorScreenshotOptions
        {
            Animations = ScreenshotAnimations.Disabled,
        }));
        var label = area + "/" + viewport;
        var screenshotPath = ScreenshotFilePath(area, viewport);
        var screenshotPresent = ScreenshotExists(area, viewport);

        if (ShouldUpdateScreenshots())
        {
            AllureApi.Step("Update screenshot: " + attachmentName, () =>
            {
                AttachUpdateMode(attachmentName, actual, screenshotPresent, area, viewport);
            });
            WriteScreenshot(screenshotPath, actual);
            return;
        }

        if (!screenshotPresent)
        {
            AllureApi.Step("Missing screenshot: " + attachmentName, () =>
            {
                AttachPng(attachmentName + "-actual-unmatched", actual);
            });
            throw new XunitException(
                $"Screenshot missing for {label}. Commit PNG to {ScreenshotResourcePath(area, viewport)} "
                + "or run with UPDATE_SCREENSHOTS=true");
        }

        var expected = ReadExpectedScreenshot(area, viewport);
        var comparison = CompareImages(expected, actual, label);
        AllureApi.Step("Compare screenshot: " + attachmentName, () =>
        {
            if (comparison.Passed)
            {
                AttachPng(attachmentName, actual);
                return;
            }

            AttachPng(attachmentName + "-expected", expected);
            AttachPng(attachmentName + "-actual", actual);
            AttachPng(attachmentName + "-diff", comparison.DiffPng);
            SaveFailArtifacts(label, actual, comparison.DiffPng);
            throw new XunitException(comparison.Message);
        });
    }

    public static string ScreenshotMode() => ScreenshotMode(ConfigReader.TestConfig.Stand);

    public static string ScreenshotMode(string? env)
    {
        var key = env?.Trim() ?? "";
        return key switch
        {
            "mock" => "mock",
            "stage" => "stage",
            "prod" or "ci" or "" => "prod",
            _ => throw new InvalidOperationException(
                $"screenshot folder: unknown env '{key}' (use mock, stage, prod, or ci)"),
        };
    }

    internal static string ScreenshotOs()
    {
        var overrideOs = Environment.GetEnvironmentVariable("SCREENSHOT_OS");
        var raw = !string.IsNullOrWhiteSpace(overrideOs) ? overrideOs.Trim() : OsFamily();
        return MapScreenshotOs(raw);
    }

    internal static string ScreenshotBrowserFolder() =>
        ScreenshotBrowser() + "-" + ScreenshotBrowserMajor();

    private static void AttachUpdateMode(string attachmentName, byte[] actual, bool screenshotPresent, string area, int viewport)
    {
        if (screenshotPresent)
        {
            AttachPng(attachmentName + "-screenshot-old", ReadExpectedScreenshot(area, viewport));
            AttachPng(attachmentName + "-screenshot-new", actual);
            return;
        }

        AttachPng(attachmentName + "-screenshot-new", actual);
    }

    private static void AttachPng(string name, byte[] png) =>
        AllureApi.AddAttachment(name, "image/png", png, ".png");

    private static bool ShouldUpdateScreenshots() => ConfigReader.TestConfig.UpdateScreenshots;

    private static string ScreenshotsDir()
    {
        var dir = ConfigReader.TestConfig.ScreenshotsDir.Trim();
        if (dir.Length == 0)
        {
            throw new InvalidOperationException("screenshotsDir must not be empty");
        }

        return dir.Replace('\\', '/').TrimEnd('/');
    }

    private static string ScreenshotBrowser()
    {
        var overrideBrowser = Environment.GetEnvironmentVariable("SCREENSHOT_BROWSER");
        return !string.IsNullOrWhiteSpace(overrideBrowser)
            ? overrideBrowser.Trim().ToLowerInvariant()
            : "chrome";
    }

    private static string ScreenshotBrowserMajor() => LocalChromePin.PinnedVersion().Split('.')[0];

    private static string OsFamily()
    {
        if (OperatingSystem.IsMacOS())
        {
            return "darwin";
        }

        if (OperatingSystem.IsWindows())
        {
            return "win32";
        }

        return "linux";
    }

    private static string MapScreenshotOs(string raw)
    {
        var key = raw.ToLowerInvariant();
        if (key is "darwin" or "macos" || key.StartsWith("mac"))
        {
            return "macos";
        }

        if (key is "win32" or "windows" || key.StartsWith("win"))
        {
            return "windows";
        }

        if (key is "linux" || key.Contains("linux"))
        {
            return "linux";
        }

        return key.Length == 0 ? "linux" : key;
    }

    private static string ScreenshotFilePath(string area, int viewport) =>
        Path.Combine(
            Config.ConfigFiles.ModuleDir(),
            ScreenshotsDir(), ScreenshotMode(), ScreenshotOs(), ScreenshotBrowserFolder(),
            area, viewport + ".png");

    private static string ScreenshotResourcePath(string area, int viewport) =>
        $"{ScreenshotsDir()}/{ScreenshotMode()}/{ScreenshotOs()}/{ScreenshotBrowserFolder()}/{area}/{viewport}.png";

    private static bool ScreenshotExists(string area, int viewport)
    {
        var output = Path.Combine(AppContext.BaseDirectory, ScreenshotResourcePath(area, viewport));
        if (File.Exists(output))
        {
            return true;
        }

        return File.Exists(ScreenshotFilePath(area, viewport));
    }

    private static byte[] ReadExpectedScreenshot(string area, int viewport)
    {
        var output = Path.Combine(AppContext.BaseDirectory, ScreenshotResourcePath(area, viewport));
        if (File.Exists(output))
        {
            return File.ReadAllBytes(output);
        }

        var path = ScreenshotFilePath(area, viewport);
        if (File.Exists(path))
        {
            return File.ReadAllBytes(path);
        }

        throw new IOException("Screenshot not found: " + ScreenshotResourcePath(area, viewport));
    }

    private static void WriteScreenshot(string screenshotPath, byte[] png)
    {
        Directory.CreateDirectory(Path.GetDirectoryName(screenshotPath)!);
        File.WriteAllBytes(screenshotPath, png);
    }

    private sealed record ImageComparison(bool Passed, byte[] DiffPng, string Message);

    private static ImageComparison CompareImages(byte[] expectedBytes, byte[] actualBytes, string label)
    {
        using var expected = Image.Load<Rgba32>(expectedBytes);
        using var actual = Image.Load<Rgba32>(actualBytes);
        var diffPng = CreateDiffPng(expected, actual);

        if (expected.Width != actual.Width || expected.Height != actual.Height)
        {
            return new ImageComparison(
                false,
                diffPng,
                $"Screenshot size changed for {label}: expected {expected.Width}x{expected.Height}, actual {actual.Width}x{actual.Height}");
        }

        var width = expected.Width;
        var height = expected.Height;
        var diffPixels = 0;
        var totalPixels = width * height;
        expected.ProcessPixelRows(actual, (expAccessor, actAccessor) =>
        {
            for (var y = 0; y < height; y++)
            {
                var expRow = expAccessor.GetRowSpan(y);
                var actRow = actAccessor.GetRowSpan(y);
                for (var x = 0; x < width; x++)
                {
                    if (expRow[x] != actRow[x])
                    {
                        diffPixels++;
                    }
                }
            }
        });

        var maxDiffRatio = ConfigReader.TestConfig.ScreenshotDiffThreshold;
        var diffRatio = (double)diffPixels / totalPixels;
        if (diffRatio > maxDiffRatio)
        {
            return new ImageComparison(
                false,
                diffPng,
                $"Screenshot diff too high for {label}: {diffRatio * 100:0.00}% > {maxDiffRatio * 100:0.00}%");
        }

        return new ImageComparison(true, diffPng, "");
    }

    private static byte[] CreateDiffPng(Image<Rgba32> expected, Image<Rgba32> actual)
    {
        var width = Math.Max(expected.Width, actual.Width);
        var height = Math.Max(expected.Height, actual.Height);
        using var diff = new Image<Rgba32>(width, height);
        for (var y = 0; y < height; y++)
        {
            for (var x = 0; x < width; x++)
            {
                var inExpected = x < expected.Width && y < expected.Height;
                var inActual = x < actual.Width && y < actual.Height;
                if (inExpected && inActual)
                {
                    var expectedRgb = expected[x, y];
                    if (expectedRgb == actual[x, y])
                    {
                        diff[x, y] = Dim(expectedRgb);
                    }
                    else
                    {
                        diff[x, y] = DiffHighlight;
                    }
                }
                else
                {
                    diff[x, y] = SizeMismatch;
                }
            }
        }

        using var ms = new MemoryStream();
        diff.SaveAsPng(ms);
        return ms.ToArray();
    }

    private static Rgba32 Dim(Rgba32 rgb)
    {
        var dim = (byte)((rgb.R + rgb.G + rgb.B) / 9);
        return new Rgba32(dim, dim, dim);
    }

    private static void SaveFailArtifacts(string label, byte[] actual, byte[] diff)
    {
        try
        {
            Directory.CreateDirectory(DiffDir);
            var prefix = label.Replace('/', '_');
            File.WriteAllBytes(Path.Combine(DiffDir, prefix + "-actual.png"), actual);
            File.WriteAllBytes(Path.Combine(DiffDir, prefix + "-diff.png"), diff);
        }
        catch (IOException)
        {
            // CI artifact is best-effort; Allure attachments are primary.
        }
    }
}
